### PR TITLE
Stochastic FSC

### PR DESCRIPTION
**Pull request by cgc**
Porting over some code I had lying around for fitting stochastic FSCs.

- [x] [Meuleau 1999 "Solving POMDPs by Searching the Space of Finite Policies"](https://arxiv.org/pdf/1301.6721.pdf)
- [ ] [Poupart 2003 "Bounded Finite State Controllers"](https://proceedings.neurips.cc/paper/2003/file/4c5bcfec8584af0d967f1ab10179ca4b-Paper.pdf)

### CODE REVIEW

File: msdm/algorithms/fscgradientascent.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+        for idx in range(self.iterations):
+            opt.zero_grad()
+
+            result = value()
+            V = result.state_value
+
+            # Check for convergence
+            if prevV is not None:
+                max_bellman_error = torch.abs(V-prevV).max().item()
```
Comment (cgc) [inline msdm/algorithms/fscgradientascent.py:132]: Flagging this for two reasons:
- a traditional convergence test through backprop seems really hard to pass with softmax-based distributions
- Doing this for state-wise values (i.e. after marginalizing out FSC state), but not quite sure if that's right.
Paper also doesn't mention convergence, so may be simplest to ax.
